File: app/src/main/java/com/yeraygarcia/recipes/database/UUIDTypeConverter.java
package com.yeraygarcia.recipes.database;

import android.arch.persistence.room.TypeConverter;

import com.fasterxml.uuid.Generators;

import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.UUID;

public class UUIDTypeConverter {

    @TypeConverter
    public static UUID toUUID(byte[] value) {
        return value == null ? null : convertBinaryToUUID(value);
    }

    @TypeConverter
    public static byte[] toByteArray(UUID value) {
        return value == null ? null : convertUUIDToBinary(value);
    }

    public static UUID newUUID() {
        return Generators.timeBasedGenerator().generate();
    }

    private static UUID convertBinaryToUUID(byte[] bytes) {
        ByteBuffer bb = reorder(bytes, false);
        long hi = bb.getLong();
        long low = bb.getLong();
        return new UUID(hi, low);
    }

    private static byte[] convertUUIDToBinary(UUID uuid) {
        byte[] bytes = new byte[16];

        // Turn into byte array
        ByteBuffer bb = ByteBuffer.allocate(16);
        bb.putLong(uuid.getMostSignificantBits());
        bb.putLong(uuid.getLeastSignificantBits());
        bb.rewind();
        bb.get(bytes);

        // Transform
        bb = reorder(bytes, true);

        // Turn into byte array
        bb.get(bytes);

        return bytes;
    }

    private static ByteBuffer reorder(byte[] bytes, boolean uuidToBin) {
        if (bytes.length != 16) {
            throw new RuntimeException("Invalid UUID bytes!");
        }

        // Get hi
        byte[] time_low;
        byte[] time_mid;
        byte[] time_hi;

        if (uuidToBin) {
            time_low = Arrays.copyOfRange(bytes, 0, 4);
            time_mid = Arrays.copyOfRange(bytes, 4, 6);
            time_hi = Arrays.copyOfRange(bytes, 6, 8); // actually time_hi + version
        } else {
            time_low = Arrays.copyOfRange(bytes, 0, 2); // actually time_hi + version
            time_mid = Arrays.copyOfRange(bytes, 2, 4);
            time_hi = Arrays.copyOfRange(bytes, 4, 8);
        }

        // Get low
        byte[] clock_seq = Arrays.copyOfRange(bytes, 8, 10); // actually variant + clock_seq
        byte[] node = Arrays.copyOfRange(bytes, 10, 16); // node

        // Rebuild
        ByteBuffer bb = ByteBuffer.allocate(16);
        bb.put(time_hi);
        bb.put(time_mid);
        bb.put(time_low);
        bb.put(clock_seq);
        bb.put(node);

        // Grab longs
        bb.rewind();
        return bb;
    }
}

File: app/src/main/java/com/yeraygarcia/recipes/database/UUIDTypeConverter.kt
package com.yeraygarcia.recipes.database

import android.arch.persistence.room.TypeConverter
import com.fasterxml.uuid.Generators
import java.nio.ByteBuffer
import java.util.*

object UUIDTypeConverter {

    @TypeConverter
    @JvmStatic
    fun toUUID(value: ByteArray?): UUID? {
        return value?.let { convertBinaryToUUID(it) }
    }

    @TypeConverter
    @JvmStatic
    fun toByteArray(value: UUID?): ByteArray? {
        return value?.let { convertUUIDToBinary(it) }
    }

    fun newUUID(): UUID {
        return Generators.timeBasedGenerator().generate()
    }

    private fun convertBinaryToUUID(bytes: ByteArray): UUID {
        val bb = reorder(bytes, false)
        val hi = bb.long
        val low = bb.long
        return UUID(hi, low)
    }

    private fun convertUUIDToBinary(uuid: UUID): ByteArray {
        val bytes = ByteArray(16)

        // Turn into byte array
        ByteBuffer.allocate(16).apply {
            putLong(uuid.mostSignificantBits)
            putLong(uuid.leastSignificantBits)
            rewind()
            get(bytes)
        }

        // Transform
        reorder(bytes, true).apply { get(bytes) }

        return bytes
    }

    private fun reorder(bytes: ByteArray, uuidToBin: Boolean): ByteBuffer {
        if (bytes.size != 16) {
            throw RuntimeException("Invalid UUID bytes!")
        }

        // Get hi
        val timeLow: ByteArray
        val timeMid: ByteArray
        val timeHigh: ByteArray

        if (uuidToBin) {
            timeLow = Arrays.copyOfRange(bytes, 0, 4)
            timeMid = Arrays.copyOfRange(bytes, 4, 6)
            timeHigh = Arrays.copyOfRange(bytes, 6, 8) // actually time_hi + version
        } else {
            timeLow = Arrays.copyOfRange(bytes, 0, 2) // actually time_hi + version
            timeMid = Arrays.copyOfRange(bytes, 2, 4)
            timeHigh = Arrays.copyOfRange(bytes, 4, 8)
        }

        // Get low
        val clockSeq = Arrays.copyOfRange(bytes, 8, 10) // actually variant + clock_seq
        val node = Arrays.copyOfRange(bytes, 10, 16) // node

        // Rebuild
        return ByteBuffer.allocate(16).apply {
            put(timeHigh)
            put(timeMid)
            put(timeLow)
            put(clockSeq)
            put(node)
            rewind()
        }
    }
}
